### ISSUE TITLE
IRB/Pry support

### ISSUE THREAD

**Issue by palkan** [ruby-next/ruby-next#80]
It would be great to have runtime transpiling support for IRB/Pry.

For example:

```sh
$ irb
> x = 1
> y = 2
> d = {x:, y:}
{x: 1, y: 2}
```

## Thoughts on implementation

It looks that the only way to do that is to monkey-patch `eval` somewhere inside the console code.

For example, IRB uses `Workspace#evaluate`: https://github.com/ruby/irb/blob/5af637b3c17f85c15a32416bc5b4579307873833/lib/irb/workspace.rb#L118

In order to activate the monkey-patch, we can use configuration files: `.irbrc` and `.pryrc`.

The open question is how to include refinements (`using RubyNext`); can we call `binding.eval("using RubyNext")` 🤔

**Comment by SteveRedka** [comment]
Does refinement even work after gems have been loaded? So far I haven't been able to even make it work in a single file: 

```ruby
require 'bundler/inline'

gemfile do
  source 'https://rubygems.org'
  gem 'ruby-next'
  gem 'pry'
end

require 'ruby-next'
require "ruby-next/language/setup"
require "ruby-next/language/runtime"

using RubyNext

x = 1
y = 2

pry
# This gives SyntaxError:
# d = {x, y}
```

**Comment by palkan** [comment]
> Does refinement even work after gems have been loaded?

Refinements are lexically scoped. For the example above, that means `using RubyNext` is only applied to the Ruby file but not runtime.

We can include refinement from the _outside_ by calling `binding.eval("using RubyNext")`; the question is how to obtain the correct binding.

**Comment by todd-a-jacobs** [comment]
## TL;DR
I don't have a solution, but I have a set of suggestions for your consideration and exploration.

## Suggested Solutions and Possibilities
1. In irb, #self is `main` (which is an instance of Object) and has a binding. I'm not sure how the internals of how RubyNext works, but wouldn't using #include or #extend on Object work? If not, maybe hooking one of the [irb commands](https://ruby-doc.org/stdlib-3.1.0/libdoc/irb/rdoc/IRB.html#module-IRB-label-Commands) would work.

1. The main problems with the posted code are that no refinement is declared, and that it's also possible that the refinement can't/won't be exported to the REPL from inside code using the refinement. A possible solution would be to allow ruby-next to monkeypatch Object rather than refining it, but that carries risks.

1. Maybe the best solution is to either refine Object in your ~/.pryrc (or ~/.irbrc), or to request that the pry/irb teams add explicit support for ruby-next.  I suggest that simply because the potential interactions between nested REPL sessions and a non-monkeypatched Object are undefined from my personal point of view.

**Comment by 4ndv** [comment]
Hi! Can you please provide a list of some edge-cases that need to be tested?

**Comment by baygeldin** [comment]
So, I've dug a little bit into it and it turns out that Pry actually copies the toplevel binding and evaluates everything in the context of the copy. That's why `using RubyNext` inside REPL throws `RuntimeError: main.using is permitted only at toplevel` even though `binding.receiver` is `main`. The key here is to enable refinments in the toplevel binding (`TOPLEVEL_BINDING.eval('using RubyNext')`) **before** Pry copies it. Luckily, Pry's loading rc files before loading REPL, so it's possible to patch the related methods in `.pryrc`. Made a proof-of-concept PR for Pry (I don't use IRB, but I bet it should be pretty similar).

In fact, it's even possible to enable both refinements and edge syntax support _right inside_ the REPL. It's a bit tricky and fragile and generally not worth it IMO, but possible. However, maybe a better solution would be to just add a couple of binaries (e.g. `pry-next` and `irb-next`) that would be exactly like original binaries, but with added `using RubyNext`. This way no patching would be needed (apart from the transpiling the code before eval).

**Comment by palkan** [comment]
Thanks @4ndv!

A couple of suggestions/questions to your snippet:
- Why do we require `"ruby-next/language/setup"` and `"ruby-next/language/runtime"`? Wouldn't `"ruby-next/language"` be enough? (And we can use `Language.transform(contents, rewriters: Language.current_rewriters, **options)` instead of `Runtime.transform`)
- We should pass `using: false` since we activate the refinement for the whole scope (TOPLEVEL_BINDING)
- Maybe we can inject `using RubyNext` within the `Workspace#evaluate` method (and use `@binding`)?
- I think, we don't need debugging here)
- Finally, I'd prefer to `prepend` + `super` (so it wouldn't break other IRB extensions)
